### ISSUE TITLE
package and branch creation/deletion workflow

### ISSUE THREAD

**Issue by celan** [aplanas/openSUSE-git#60]
Manage CRUD of packages in maintained products/projects, in particular the workflow of how to add or delete a package from an existing product with git methods.

**Comment by celan** [comment]
@dirkmueller @adrianschroeter @AdamMajer 
Do I understand this correctly that this basically refers to a facility to add newly to be maintained packages to products in development or Maintenance?

I think this issue could benefit from more words or re-phrasing, because it is a bit misleading, because of its brevity.

**Comment by dirkmueller** [comment]
yes, in particular the workflow of how to add or delete a package from an existing product with git methods

**Comment by mcepl** [comment]
> […] from an existing product […]

??? I thought it should be something like `osc mkpac`, but this sounds more like `osc branch` or `osc copypac` (which would be a duplicate of #56). ???

**Comment by AdamMajer** [comment]
This is about adding new packages to and removing packages from a Git Project via PR workflow. It's basically needs a tool that will generate a Git Project PR **and** maybe change ownership of the package in Gitea.

We require all packages to be owned by `/pool`  before they can be part of any openSUSE/SUSE official project. So adding a new package from a devel project to openSUSE:Factory for the first time would be,

  1. change ownership from devel project to /pool
  2. create a fork in devel project (so devel project continues building)
  3. create PR against Factory project

Removing a package from a project git is simpler. On accept, tag removed-<branch>-<PR#> (so older commits are not GC'ed) and delete the branch. But still would need a helper to automate the removal PR creation.

**Comment by mcepl** [comment]
Just beginning of the code is at https://git.sr.ht/~mcepl/gosc/tree/mkpac/item/git-mkpac.sh. Currently my question is what we could do with authorization:

* tea repo create is convenient, but it cannot create sha256 repositories ([patch for tea?](https://gitea.com/gitea/tea/issues/727))
* we can get token via

```sh
  curl -H "Content-Type: application/json" -d '{"name":"test"}' -u ${login}:${password} \
    https://src.opensuse.org/api/v1/users/${login}/tokens
```

but we don’t use login password, how does this curl command can be made working with https://id.opensuse.org?

  Then we could do

    curl -X POST 'https://src.opensuse.org/api/v1/user/repos' \
      -H 'accept: application/json' \
      -H "Authorization: token $TOKEN" \
      -H 'Content-Type: application/json' \
      -d '{
        "default_branch": "factory",
        "description": "Wayland clipboard utility",
        "name": "wayclip1",
        "object_format_name": "sha256"
      }'

but how to do the authorization?

**Comment by JanZerebecki** [comment]
I don't know if it is sufficiently secure, but how about using Oath for authorizing the gitea requests?: https://docs.gitea.com/next/development/oauth2-provider

**Comment by bmwiedemann** [comment]
You can create a token manually at https://src.opensuse.org/user/settings/applications

**Comment by mcepl** [comment]
Filed https://gitea.com/gitea/tea/issues/727 for creating new repo with `object-format=sha256`.

**Comment by mcepl** [comment]
BTW, `tea repo create-from-template --name test --template pool/new_package` works correctly and creates SHA256 repository.

**Comment by mmaslano** [comment]
@mcepl do you feel it's possible to remove or add packages now?

**Comment by hramrach** [comment]
From the description it is `osc sr` for the case when the target does not exist, not mkpac.

For that case no usable solution exists.

The solution provided so far is 'ask maintainer of target project to manually create new package'.

Deletes should be possible be creating a PR that removes the package from the project (without any additional cleanup).

**Comment by hramrach** [comment]
There is some 'transfer repository' feature in gitea that might potentially be usable but it probably transfers too much. 

There is 'reverse fork' feature which would likely do exactly the needed thing but it is not implemented  https://github.com/go-gitea/gitea/issues/34848

**Comment by AdamMajer** [comment]
We have that patch currently. It may need reworking for final version.

**Comment by hramrach** [comment]
Also this is not limited to creating repositories. When the repository exists but does not have the required branch there is a similar problem. No way to open a PR  AFAIK.

**Comment by mcepl** [comment]
> [@mcepl](https://github.com/mcepl) do you feel it's possible to remove or add packages now?

What @hramrach wrote: workflow “send email to somebody crazily overloaded with work and hope they would eventually create a new package/branch for you” is not sustainable and doesn’t scale at all.

We have now a good way (via Web UI or `tee` command) to create new repo in `/$USER` namespace, but no good way how to get it to `/pool`.

So, no, as of now, I don’t think this ticket has been resolved.

**Comment by mmaslano** [comment]
I'm closing this. Adding/removal of packages was replaced by TW requirements for addition/removal of packages.

**Comment by hramrach** [comment]
See #146
